### PR TITLE
Implement string comparison

### LINKED ISSUE #197: String comparison
Currently string comparison such as `assert x = "abc"` doesn't work properly yet, I think we should implement the operation in LPython in the runtime library and the frontend should insert the call to it.

### PR DESCRIPTION
**Pull request by Abdelrahman-Kh-Fouad**
Fixes #197 

with basic testing `a == "sd"` it give me `Stored value type does not match pointer operand type!`, how to fix return type to be bool? or Is the problem something else?
```bash
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store i8* %4, i1* %s, align 8
 i1
```
@namannimmo10

### CODE REVIEW

**Comment by Abdelrahman-Kh-Fouad** [comment]
> You are using `lfortran_strop` to call into C, but it assumes that the third argument is of type `char**`. See `lfortran_strrepeat` for example, it assumes the second argument is of type `int32_t`. So I think you need a different function (maybe `lfortran_strcmp`) to handle string comparison.

i made `lfortran_str_cmp` to accept `int8` as a result and it didn't work, i tried to make functions return value like `str_len` (last commit) and it didn't work either and, they gave me same message 
```bash
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store i8 %3, i1* %s, align 1
 i1
```
is `i1` refere to void?

**Comment by namannimmo10** [comment]
`i1` is a boolean value (`True` or `False`), a 1-bit integer.

**Comment by namannimmo10** [comment]
This looks better now. Please add the tests covering all corner cases and test them with `cpython` and `llvm`.

**Comment by namannimmo10** [comment]
You also need to add this file in https://github.com/lcompilers/lpython/blob/main/integration_tests/CMakeLists.txt with labels, `cpython` and `llvm`.

**Comment by namannimmo10** [comment]
Please do a `git rebase` with the `main` branch rather than merge it with `main`. It leads to unnecessary commits, making a mess in the history. 
```console
$ git checkout main
$ git pull upstream main
$ git checkout stringsCompare
$ git rebase main   # if ever this leads to conflicts - resolve manually
```

**Comment by Abdelrahman-Kh-Fouad** [comment]
> Please do a `git rebase` with the `main` branch rather than merge it with `main`. It leads to unnecessary commits, making a mess in the history.
> 
> ```
> $ git checkout main
> $ git pull upstream main
> $ git checkout stringsCompare
> $ git rebase main   # if ever this leads to conflicts - resolve manually
> ```

i rebased the branch as you said but, it didn't squash the branch.

**Comment by namannimmo10** [comment]
You don't have to squash all the commits into one. Do `git rebase -i main` and it will show you the commits in an interactive mode. Now you can change `pick` to `s` to squash the commit, `d` to drop, and so on... For instance, you have three commits with the message, "Solve conflicts" that are not needed here, so change `pick` to `s`, so that we only have a relevant set of commits. When you're done, close the editor, and push it with `--force`.

File: tests/tests.toml [hunk 1]
```diff
@@ -346,6 +346,12 @@ filename = "../integration_tests/print_01.py"
 c = true
 cpp = true
 
+[[test]]
+filename = "../integration_tests/test_str_comparison.py"
+llvm = true
+c = true
+cpp = true 
+
```
Comment (namannimmo10) [inline tests/tests.toml:354]:
```suggestion
```
These are not needed. Just testing with `cpython` and `llvm` should be enough, I think.

Comment (Abdelrahman-Kh-Fouad) [reply tests/tests.toml]: I deleted it.

**Comment by namannimmo10** [comment]
You also need to delete those 6 generated files.

File: integration_tests/test_str_comparison.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+def f():
+    s1: str = "abcd"
+    s2: str = "abcd"
+    assert s1 == s2
+    assert s1 <= s2
+    assert s1 >= s2
+    s1 = "abcde"
+    assert s1 >= s2
```
Comment (namannimmo10) [inline integration_tests/test_str_comparison.py:8]: A test for the `!=` operator would be nice too!

File: src/runtime/impure/lfortran_intrinsics.c [hunk 3]
```diff
@@ -614,6 +614,53 @@ LFORTRAN_API void _lfortran_strcat(char** s1, char** s2, char** dest)
     *dest = &(dest_char[0]);
 }
 
+#define MIN(x, y) (((x) < (y)) ? (x) : (y))
```
Comment (namannimmo10) [inline src/runtime/impure/lfortran_intrinsics.c:617]:
```suggestion
#define MIN(x, y) ((x < y) ? x : y)
```

**Comment by namannimmo10** [comment]
Please make these changes and it should be ready afterward.

File: integration_tests/out [hunk 4]
```diff
@@ -0,0 +1,696 @@
+-- The C compiler identification is GNU 11.2.0
```
Comment (certik) [inline integration_tests/out:1]: I think this file should not be committed.

**Comment by certik** [comment]
After everything is ready, let's create a simple commit or commits and rebased on top of the latest master.

**Comment by Abdelrahman-Kh-Fouad** [comment]
> After everything is ready, let's create a simple commit or commits and rebased on top of the latest master.

it's ready now

**Review by certik — APPROVED** [review]
That looks good to me, I think we can merge it.

Would you mind creating a nice set of patches for this? Or do you want me to just squash it using github?

**Comment by Abdelrahman-Kh-Fouad** [comment]
> Would you mind creating a nice set of patches for

i don't mind but, how to create a set of patches?

**Comment by namannimmo10** [comment]
I squashed all the commits into one for now. @Abdelrahman-Kh-Fouad -- thanks a lot for working on this! If you still want to know how to squash commits, we can show you in our upcoming meeting.
